### ISSUE TITLE
Transaction isolation?

### ISSUE THREAD

**Issue by kevinburkeshyp** [brianc/node-postgres#794]
Maybe I am confused about how database clients work... I would appreciate some clarity on this. I understand a Node process can respond to two requests simultaneously, yielding via nextTick, etc.

What happens if both requests want to start a transaction, how does node-postgres/the database keep them separate? E.g. if I have two BEGINs coming from the same process, and then issue a rollback, how does Postgres know which transaction to rollback?

**Comment by brianc** [comment]
As far as postgres is concerned it doesn't care if you're sending in 2 queries "at the same time" from a node app via async io or you're sending 2 queries in "at the same time" from a Java app w/ threads on the server or from a single threaded rails app but 2 different instances behind a load balancer - it all looks the same to postgres.

Postgres manages concurrency on a per-connection basis.  If you have two clients (and therefore two connections) and one wants to start a transaction and locks a particular row in the database postgres will do what it needs to to ensure isolation internally for the other connections, both existing and those established before the locked resource is released.  

Node postgres's role here is pretty minimal - it delegates all proper responsibilities to the database server. It mostly parses the protocol, turns raw bytes into javascript messages, and does a bit of connection pooling.

**Comment by brianc** [comment]
tl; dr - node-postgres works pretty much the same as a postgres client driver in any other language or platform you're familiar with.  node-postgres opens and manages a connection for each `new Client` instance you create. The isolation level is guaranteed on a per-client level by postgres itself.

Hope this helps! :smile:

**Comment by kevinburkeshyp** [comment]
If a process (say, an Express web server) is handling two HTTP requests is it possible that they'd share the same database session? or would I have to check the implementation of the server/client pool?

Request 1 issues BEGIN, yields; 
Request 2 issues BEGIN, yields;
Request 1 tries to update record 1 to value A
Request 2 tries to update record 1 to value B
Either issues a COMMIT or a ROLLBACK.

**Comment by brianc** [comment]
I wrote this a while back - does this help?

https://gist.github.com/brianc/f906bacc17409203aee0

**Comment by kevinburkeshyp** [comment]
it does - thanks!

**Comment by kevinburkeshyp** [comment]
just a heads up - that gist could probably use some logic for rollbacks, right? if the query doesn't succeed or whatever..

**Comment by noguilty4you** [comment]
It is much simpler than I thought

`
const DBClient = await new Pg.Pool(<config>).connect()

await DBClient.query('BEGIN')

await DBClient.query('SET TRANSACTION ISOLATION LEVEL REPEATABLE READ')

// ...payload

`
